### PR TITLE
Progressive log text

### PR DESCRIPTION
**Pull request by timbrown5**
Fix for: #175

Changes:
- [x] PipelineConsole polls for status updates.
- [x] Changed StepInfo.id from number to string (it was converted to string in many places).
- [x] Split log text from StepInfo object (allows states to be updated independently without remembering to copy values over).
- [x] Follow new steps when at bottom of the step list (and not following the log?).
- [x] (to fix) Expanded steps stay expanded when the state gets updated.

Future work:
- [ ] Log following when at bottom of step log.
-- The list following built into Virtuoso isn't working for this so would need to investigate why. I think it's worth doing this in a separate Issue/PR to allow progressive logs to land - but with the annoyance of needing to scroll. That said if anyone knows how to fix it that would be great 😅.

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/frontend/pipeline-console-view/pipeline-console/main/PipelineConsole.tsx [hunk 1]
```diff
@@ -289,9 +350,13 @@ export default class PipelineConsole extends React.Component<
     this.updateStepConsoleOffset(nodeId, true, startByte);
   }
 
+  //handleFollowLog(nodeId: string) {
```
Comment (timja) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/PipelineConsole.tsx:353]: ?

File: src/main/frontend/pipeline-console-view/pipeline-console/main/StageView.tsx [hunk 2]
```diff
@@ -142,19 +143,18 @@ export default class StageView extends React.Component {
...
-          String(stepItemData.id)
-        )}`
-      );
+    return stepsItems.map((stepItemData, index) => {
       return (
         <ConsoleLogCard
           step={stepItemData}
+          stepBuffer={this.props.stepBuffers.get(stepItemData.id) ?? {
+            consoleLines: [] as string[],
+            consoleStartByte: - LOG_FETCH_SIZE,
```
Comment (timja) [inline src/main/frontend/pipeline-console-view/pipeline-console/main/StageView.tsx:152]: this looks a bit odd, is it a shorthand for `0 - LOG_FETCH_SIZE`?

**Review by timja — COMMENTED** [review]
Seems good for a first version.

One other issue is that the build status icon doesn't update when the build finishes it just keeps spinning

File: src/main/frontend/pipeline-console-view/pipeline-console/main/PipelineConsole.tsx [hunk 3]
```diff
@@ -44,17 +49,47 @@ export default class PipelineConsole extends React.Component<
       expandedSteps: [] as string[],
       stages: [] as StageInfo[],
       steps: [] as StepInfo[],
+      // Use string as key so we don't need to look through steps when we update console log.
+      // Should really make StepInfo id a string.
```
[the remark this answers is not in the document]
Comment (timbrown5) [reply src/main/frontend/pipeline-console-view/pipeline-console/main/PipelineConsole.tsx:53]: Yes, just after I wrote that comment 🙄

**Review by timja — APPROVED** [review]

**Comment by timbrown5** [comment]
That's true, and I think assumed it was tracked by #239 - my guess is the issue is cause by the Jelly re;reading the value of `it.buildStatusIconClassName` here:
https://github.com/jenkinsci/pipeline-graph-view-plugin/blob/main/src/main/resources/io/jenkins/plugins/pipelinegraphview/consoleview/PipelineConsoleViewAction/index.jelly#L10
I could try forcing a page refresh, but I think all the console text would be reset :/.

**Pull request merged**